File: profile/time/primes.py
import argparse
import logging
import math
import sys
import time

from collections.abc import Generator


def sieve(n : int) -> list[int]:
    '''Yield the prime numbers between 2 and n.'''
    arr: list[bool] = [True for _ in range(n+1)]

    for i in range(2, math.ceil(math.sqrt(n)) + 1):
        if arr[i]:
            for j in range(i*i, n+1, i):
                arr[j] = False

    return list(i for i, is_prime in enumerate(arr[2:n+1], 2) if is_prime)
            


def main():
    parser = argparse.ArgumentParser(
        description='Print the largest prime numbers no greater than n.')
    parser.add_argument('n', help='maximum number to print', type=int)
    args = parser.parse_args()
    
    # start = time.monotonic()
    primes = sieve(args.n)
    # end = time.monotonic()
    print(primes[-1])
    
    # print("calculating took %.2f seconds" % (end - start,))


if __name__ == '__main__':
    main()

File: profile/time/primes.cpp
#include <chrono>
#include <cmath>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>


std::vector<long> sieve(long n) {
    std::vector<bool> arr((unsigned long)(n+1), true);

    for (long i = 2; i <= (long)std::ceil(std::sqrt(n)); ++i) {
        if (arr[i]) {
            for (long j = i*i; j <= n; j += i) {
                arr[j] = false;
            }
        }
    }

    std::vector<long> result;
    for (long k = 2; k <= n; ++k) {
        if (arr[k]) {
            result.emplace_back(k);
        }
    }
    return result;
}


[[noreturn]] void fail(const std::string& message, char* arg)
{
    std::cerr << message << ": '" << arg << "'" << std::endl;
    std::exit(EXIT_FAILURE);
}


int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "usage: " << argv[0] << " n" << std::endl;
        return EXIT_FAILURE; 
    }

    long n;
    try {
        n = std::stol(argv[1]);
    }
    catch (std::invalid_argument const& ex) {
        fail("invalid integer value", argv[1]);
    }
    catch (std::out_of_range const& ex) {
        fail("integer value out of range", argv[1]);
    }

    if (n < 2) {
        fail("argument must not be less than two", argv[1]);
    }

    // const auto start{std::chrono::steady_clock::now()};
    std::vector<long> primes = sieve(n);
    // const auto end{std::chrono::steady_clock::now()};

    std::cout << primes[primes.size() - 1] << "\n";

    // const std::chrono::duration<double> elapsed{end - start};
    // std::cout << "calculating took "
    //           << std::fixed << std::setprecision(2) << elapsed.count()
    //           << " seconds" << std::endl;

    return EXIT_SUCCESS;
}
